### PR TITLE
Remove batch manager with feature gate

### PR DESCRIPTION
**Pull request by andre4i**
See #3788

### CODE REVIEW

File: BREAKING.md [hunk 1]
```diff
@@ -4,6 +4,8 @@
 - [documentId removed from IFluidDataStoreRuntime and IFluidDataStoreContext](#documentId-removed-from-IFluidDataStoreRuntime-and-IFluidDataStoreContext)
 - [@fluid-experimental/tinylicious-client package name changed](#fluid-experimentaltinylicious-client-package-name-changed)
 - [@fluid-experimental/fluid-static package name changed](#fluid-experimentalfluid-static-package-name-changed)
+- [FlushMode enum values renamed](#FlushMode-enum-values-renamed)
```
Comment (vladsud) [inline BREAKING.md:7]: I think it's great to have it as a single change to experiment, but I'd suggest submitting it in staging:
1. Change in naming (simple, unless we find it crosses compat layers)
2. Change in default batching behavior
3. Removal of batch manager.

Comment (andre4i) [reply BREAKING.md:7]: 1&2 are in https://github.com/microsoft/FluidFramework/pull/7071, although I see it breaks a test (starting today after merging in main).

File: BREAKING.md [hunk 2]
```diff
@@ -29,6 +31,14 @@ The `@fluid-experimental/tinylicious-client` package has been renamed to now be
 ### `@fluid-experimental/fluid-static` package name changed
 The `@fluid-experimental/fluid-static` package has been renamed to now be `@fluidframework/fluid-static`.
 
+### `FlushMode` enum values renamed
+`FlushMode` enum values from `@fluidframework/runtime-definitions` have ben renamed as following:
+- `FlushMode.Manual` to `FlushMode.TurnBased`
+- `FlushMode.Automatic` to `FlushMode.Immediate`
+
+### `FlushMode` default changed for `ContainerRuntime`
+Previously, `ContainerRuntime` had the default flush mode as `FlushMode.Immediate` (formerly `FlushMode.Automatic`). This default has been changed to `FlushMode.TurnBased` (formerly `FlushMode.Manual`). Note that the `batchBegin` and `batchEnd` events from `IContainerRuntime` will no longer be emitted for each processed op by default.
```
Comment (vladsud) [inline BREAKING.md:40]: It will be hard for people to consume that paragraph, thus I'd separate (across couple versions) rename vs. change in behavior.

`batchBegin` and `batchEnd` - This reads scary, as summarizer depends on it, to be able to inject summary at random time. So the first time I read it, I felt like everything will break :).
I think we want to soften message and say that natural batch granularity is extended because of these changes, i.e. batch scope used to be (on average) 1 op vs. now it's likely to be larger (scenario dependent).

File: api-report/runtime-definitions.api.md [hunk 3]
```diff
@@ -62,8 +62,8 @@ export type FluidDataStoreRegistryEntry = Readonly<Partial<IProvideFluidDataStor
 
 // @public
 export enum FlushMode {
-    Automatic = 0,
-    Manual = 1
+    Immediate = 0,
```
Comment (vladsud) [inline api-report/runtime-definitions.api.md:65]: So, given these are numbers, this is not breaking change, and compat UTs should pass, right? Just confirming.

File: BREAKING.md (hunk 2 above)
Comment (andre4i) [reply BREAKING.md:40]: Thanks for the feedback, will redo. Can you look at https://github.com/microsoft/FluidFramework/pull/7071 where these changes are staged for merging? They are brought here in this PR only so that I can move forward with the batchmanager work.

File: packages/drivers/local-driver/src/localDocumentDeltaConnection.ts [hunk 4]
```diff
@@ -69,17 +69,15 @@ export class LocalDocumentDeltaConnection extends DocumentDeltaConnection {
         // We use a promise resolve to force a turn break given message processing is sync
         // eslint-disable-next-line @typescript-eslint/no-floating-promises
         Promise.resolve().then(() => {
-            this.submitManager.add("submitOp", messages);
-            this.submitManager.drain();
+            this.submitCore("submitOp", messages);
```
Comment (vladsud) [inline packages/drivers/local-driver/src/localDocumentDeltaConnection.ts:72]: We have this problem without your changes, but it will become worse - Kafka limit for a batch is 1M. So if payload exceeds that size, it fails. I think driver has to have some hard limit that is lower that results in critical (non-retriable failure) and runtime should have the warning of half of that size (even though I believe limit is not communicated to runtime). That said it would be also great to propagate that limit from service to driver to runtime, in similar way we propagate service policy RE summarization config. Please take a look (or open an issue to follow up), I can dig examples of how it's done in summarization space.

File: BREAKING.md (hunk 1 above)
Comment (vladsud) [reply BREAKING.md:7]: Also I'd suggest flighting # 2. I.e. ship it dark and have some overwrite (localStorage for lack of better tools) to enable it in Bohemia and see what breaks. We (I :)) might be underestimating impact and the need to turn it off to give people enough time to address bugs that likely were always there but were hidden / less likely to be hit.

Comment (andre4i) [reply BREAKING.md:7]: I agree with this approach and will do that. First I need to merge https://github.com/microsoft/FluidFramework/pull/7071 after I figure out why the `todo item` tests started failing intermittently. This PR also saw the test failures but they cleared after re-running. I want to first make sure I'm not adding instability with the flushmode change..

File: api-report/runtime-definitions.api.md (hunk 3 above)
Comment (andre4i) [reply api-report/runtime-definitions.api.md:65]: Yes. It works :)

File: packages/drivers/local-driver/src/localDocumentDeltaConnection.ts (hunk 4 above)
Comment (andre4i) [reply packages/drivers/local-driver/src/localDocumentDeltaConnection.ts:81]: https://github.com/microsoft/FluidFramework/issues/7246

File: packages/drivers/local-driver/src/localDocumentDeltaConnection.ts [hunk 5]
```diff
@@ -62,24 +62,31 @@ export class LocalDocumentDeltaConnection extends DocumentDeltaConnection {
           super(socket, documentId, new TelemetryNullLogger());
     }
 
+    protected submitCore(type: string, messages: IDocumentMessage[]) {
+        if (this.isBatchManagerDisabled) {
+            this.emitMessages(type, [messages]);
+        } else {
+            this.submitManager.add(type, messages);
+            this.submitManager.drain();
```
Comment (vladsud) [inline packages/drivers/local-driver/src/localDocumentDeltaConnection.ts:70]: It's not very clear why this dance of add + drain is required. Can we drop this implementation of submitCore in favor of parent's implementation?

File: packages/drivers/driver-base/src/documentDeltaConnection.ts [hunk 6]
```diff
@@ -259,14 +258,40 @@ export class DocumentDeltaConnection
...
+
     /**
      * Submits a new delta operation to the server
      *
      * @param message - delta operation to submit
      */
     public submit(messages: IDocumentMessage[]): void {
         this.checkNotClosed();
-        this.submitManager.add("submitOp", messages);
+        this.submitCore("submitOp", messages);
```
Comment (vladsud) [inline packages/drivers/driver-base/src/documentDeltaConnection.ts:294]: I think it should be [messages] in here

Comment (andre4i) [reply packages/drivers/driver-base/src/documentDeltaConnection.ts:294]: `submitCore` already does that

File: packages/drivers/driver-base/src/documentDeltaConnection.ts [hunk 7]
```diff
@@ -259,14 +258,40 @@ export class DocumentDeltaConnection
...
+        }
+    }
+
     /**
      * Submits a new delta operation to the server
      *
      * @param message - delta operation to submit
      */
     public submit(messages: IDocumentMessage[]): void {
         this.checkNotClosed();
```
Comment (vladsud) [inline packages/drivers/driver-base/src/documentDeltaConnection.ts:293]: Parent checks for not closed, so I think all the methods in this class can drop checking it.
And I think parent class should have same semantics as this class, i.e. if batch manager is disabled, then it should assert if any submission is attempted while closed. The reason it was not that type of check was because of batching / delayed processing of ops (with batch manager in place), which introduced race condition, i.e. it was normal for draining to occur after closure. With synchnornos flow it's a violation of contract.

File: packages/drivers/local-driver/src/localDocumentDeltaConnection.ts (hunk 5 above)
Comment (andre4i) [reply packages/drivers/local-driver/src/localDocumentDeltaConnection.ts:70]: Not sure I follow. This is to keep the existing feature (batchmanager) as is, so I kept the original code here..

File: packages/drivers/local-driver/src/localDocumentDeltaConnection.ts [hunk 8]
```diff
@@ -62,24 +62,31 @@ export class LocalDocumentDeltaConnection extends DocumentDeltaConnection {
...
     /**
      * Submits a new delta operation to the server
      */
     public submit(messages: IDocumentMessage[]): void {
         // We use a promise resolve to force a turn break given message processing is sync
         // eslint-disable-next-line @typescript-eslint/no-floating-promises
         Promise.resolve().then(() => {
-            this.submitManager.add("submitOp", messages);
-            this.submitManager.drain();
+            this.submitCore("submitOp", messages);
```
Comment (vladsud) [inline packages/drivers/local-driver/src/localDocumentDeltaConnection.ts:81]: [messages] ?
Just to make sure that contract is clear: it's an array of batches, where batch can be either single op or many ops. So submitting just an array is valid, but it breaks contract - it no longer a single batch.
And while it's valid RE protocol, I think we should disallow sending array of ops. It should be array of arrays always, where inner array can have just one op. That way we should be able to have proper type guards in this code and catch cases like this where we are passing just an array.

Another solution - stop supporting multiple batches in these layers (as input). I.e. multiple batches being sent at once can only happen due to BatchManager, so it feels like all APIs before should just take an array (single batch), and there should be no need to have [messages[ in other places

Comment (vladsud) [reply packages/drivers/local-driver/src/localDocumentDeltaConnection.ts:81]: Curious, is there a way to write UT that will validate that messages on socket show up in right form (i.e. batch is preserved)?

File: packages/drivers/local-driver/src/localDocumentDeltaConnection.ts (hunk 5 above)
Comment (vladsud) [reply packages/drivers/local-driver/src/localDocumentDeltaConnection.ts:70]: Can this function be deleted, and thus parent submitCore() used? Will it work? Basically do we need this code?

File: packages/drivers/driver-base/src/documentDeltaConnection.ts (hunk 6 above)
Comment (vladsud) [reply packages/drivers/driver-base/src/documentDeltaConnection.ts:294]: Ah, you are right. It's just hard to follow as some functions take an array and some do not (as their input)

File: packages/drivers/local-driver/src/localDocumentDeltaConnection.ts (hunk 8 above)
Comment (vladsud) [reply packages/drivers/local-driver/src/localDocumentDeltaConnection.ts:81]: Per other comment, this seems correct, but question about UT might be still relevant (though TSC properly validates types, then it's less of an issue)

**Review by vladsud — APPROVED** [review]

File: packages/drivers/local-driver/src/localDocumentDeltaConnection.ts (hunk 8 above)
Comment (andre4i) [reply packages/drivers/local-driver/src/localDocumentDeltaConnection.ts:81]: The submitCore implementation takes care of that. It has the same exact signature as `batchmanager.add`. There's no regression here.

File: packages/drivers/local-driver/src/localDocumentDeltaConnection.ts (hunk 5 above)
Comment (andre4i) [reply packages/drivers/local-driver/src/localDocumentDeltaConnection.ts:70]: > Basically do we need this code?

I'm not sure and at this point I'm afraid to ask :))

Let me do some experimenting. I'd rather NOT do this, you have a point.

File: packages/drivers/driver-base/src/documentDeltaConnection.ts (hunk 7 above)
Comment (andre4i) [reply packages/drivers/driver-base/src/documentDeltaConnection.ts:293]: > Parent checks for not closed

Can you clarify which parent are you referring to?

File: packages/drivers/local-driver/src/localDocumentDeltaConnection.ts (hunk 5 above)
Comment (andre4i) [reply packages/drivers/local-driver/src/localDocumentDeltaConnection.ts:70]: There are some DDS tests failing consistently if these ops are not drained. Considering the code will be (eventually) removed and the `emit` above is basically `submitManager.drain` with fewer steps, I'll leave this as is.
